src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs

```diff
@@ -839,6 +839,181 @@ internal void OnEnvChange(SqlEnvChange rec)
+                         || ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryWorkloadIdentity
+                         || (ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryIntegrated && _fedAuthRequired),
+                         "Credentials aren't provided for calling MSAL");
+            Debug.Assert(fedAuthInfo != null, "info should not be null.");
+            Debug.Assert(_dbConnectionPoolAuthenticationContextKey == null,
+                "_dbConnectionPoolAuthenticationContextKey should be null.");
+            SqlClientEventSource.Log.TryTraceEvent("<sc.SqlInternalConnectionTds.OnFedAuthInfo> {0}, Generating federated authentication token", ObjectID);
+            DbConnectionPoolAuthenticationContext dbConnectionPoolAuthenticationContext = null;
+
+            // We want to refresh the token without taking a lock on the context, allowed when the
+            // access token is expiring within the next 10 mins.
+            bool attemptRefreshTokenUnLocked = false;
+
+            // We want to refresh the token, if taking the lock on the authentication context is
+            // successful.
+            bool attemptRefreshTokenLocked = false;
+
+            if (_dbConnectionPool != null)
+            {
+                Debug.Assert(_dbConnectionPool.AuthenticationContexts != null);
+
+                // Construct the dbAuthenticationContextKey with information from FedAuthInfo and
+                // store for later use, when inserting in to the token cache.
+                _dbConnectionPoolAuthenticationContextKey = new DbConnectionPoolAuthenticationContextKey(
+                    fedAuthInfo.stsurl,
+                    fedAuthInfo.spn);
+
+                // Try to retrieve the authentication context from the pool, if one does exist for
+                // this key.
+                if (_dbConnectionPool.AuthenticationContexts.TryGetValue(_dbConnectionPoolAuthenticationContextKey, out dbConnectionPoolAuthenticationContext))
+                {
+                    Debug.Assert(dbConnectionPoolAuthenticationContext != null,
+                        "dbConnectionPoolAuthenticationContext should not be null.");
+
+                    // The timespan between UTCNow and the token expiry.
+                    TimeSpan contextValidity =
+                        dbConnectionPoolAuthenticationContext.ExpirationTime.Subtract(DateTime.UtcNow);
+
+
+                    if (contextValidity <= _dbAuthenticationContextUnLockedRefreshTimeSpan)
+                    {
+                        // If the authentication context is expiring within next 10 minutes, lets
+                        // just re-create a token for this connection attempt. And on successful
+                        // login, try to update the cache with the new token.
+                        SqlClientEventSource.Log.TryTraceEvent(
```

Comment:
Removed if trace enabled check and replaced with version that already does that check